### PR TITLE
Sort options by file time and by most sung

### PR DESCRIPTION
**Pull request by twollgam**
### What does this PR do?

The pr adds new song sort options:
- by file time
- by most sung
 
Additionally it stores date and time of performs and display them.

### Closes Issue(s)

#787 and #822 

### Motivation

I am missing those options during game play :-) Additionally storing dates will be used for player achievements later. 

### More

- [x] Needs testing
- [x] Needs cut off for displaying scores after singing cause now 16 scores are stored per song. Before only 3 scores were stored and displayed. (Draw 8 scores/lines)
- [x] Draw score dates.
- [x] Fix bug in saving option_list

### Additional Notes

None

### CODE REVIEW

**Comment by twollgam** [comment]
@ooshlablu could you test this pr? Thanks in advance :heart:

**Comment by ooshlablu** [comment]
I tested this with the deb and AppImage on Ubuntu 20.04 & 22.04. Both options seem to work, and I have not noticed any detrimental effects to existing functionality or game play. LGTM

**Comment by twollgam** [comment]
@ooshlablu would be nice if you would test this pr.
- In Configure->Game->Date time format (after Language) one can select a format that is used to display the date and time when a song was performed on the song screen. The selected format should be saved and restored after restart of the game.
- Every song that is performed and enters the highscore list of this song should be displayed with the date time.
- The number of scores displayed is 8.
- The number of scores stored in the database should be 16 IIRC.
- Sorting option "Most sung" works with the number of scores.
- Sorting option "File time" works with last write time of the song folder.

**Comment by twollgam** [comment]
@Baklap4 @OznOg  would be nice if someone would do a code review. TIA

**Comment by Baklap4** [comment]
@twollgam end of next week i'll have some time going through it
Just went through it very roughly and saw you extracted the sorting options all together which is nice! clean short classes i love them.

File: game/configuration.cc
Comment (Baklap4) [inline game/configuration.cc:231]: please be consistent, always add braces

File: game/configuration.cc
Comment (Baklap4) [inline game/configuration.cc:278]: please be consistent, always add braces

File: game/database.cc
Comment (Baklap4) [inline game/database.cc:98]: Braces

File: game/database.cc
Comment (Baklap4) [inline game/database.cc:119]: Is there a need to log this message?

File: game/database.cc
Comment (Baklap4) [inline game/database.cc:110]: Is there a need for logging this message?

File: game/hiscore.cc
Comment (Baklap4) [inline game/hiscore.cc:23]: magic number, please make a constant for this. For now it looks alright, but once someone starts to refactor it might not be alright anymore

File: game/hiscore.cc
Comment (Baklap4) [inline game/hiscore.cc:14]: magic number, please make a constant for this. For now it looks alright, but once someone starts to refactor it might not be alright anymore

File: game/hiscore.cc
Comment (Baklap4) [inline game/hiscore.cc:62]: Braces

File: game/hiscore.cc
Comment (Baklap4) [inline game/hiscore.cc:71]: Braces

File: game/hiscore.cc
Comment (Baklap4) [inline game/hiscore.cc:81]: Braces

File: game/hiscore.cc
Comment (Baklap4) [inline game/hiscore.cc:84]: Braces

File: game/hiscore.cc
Comment (Baklap4) [inline game/hiscore.cc:97]: Braces

File: game/hiscore.cc
Comment (Baklap4) [inline game/hiscore.cc:102]: Braces

File: game/screen_songs.cc
Comment (Baklap4) [inline game/screen_songs.cc:368]: Braces

File: game/songs.cc
Comment (Baklap4) [inline game/songs.cc:502]: Braces

File: testing/utiltest.cc
Comment (Baklap4) [inline testing/utiltest.cc:39]: I think we can remove this comment, as the actual used are not using those compound things

File: .github/workflows/linux.yml
Comment (twollgam) [reply .github/workflows/linux.yml:1]: It it a workaround suggested by @ooshlablu. Can this be removed?

File: game/database.cc
Comment (twollgam) [reply game/database.cc:110]: No, not here, cause this is a request, if the highscore already exist. The exception is not an error in this case.

File: game/database.cc
Comment (twollgam) [reply game/database.cc:119]: I  am not sure. I think it is not needed and we should not add it yet.

File: game/hiscore.cc
Comment (twollgam) [reply game/hiscore.cc:14]: I added constants for both numbers.

File: testing/utiltest.cc
Comment (twollgam) [reply testing/utiltest.cc:39]: I think we should keep that, even if I don't like comments, but here is a bug in a foreign component which is not obvious.

File: .github/workflows/linux.yml
Comment (twollgam) [reply .github/workflows/linux.yml]: Removed the workaround.

File: game/hiscore.cc
Comment (twollgam) [reply game/hiscore.cc]: @Baklap4 please take a look to the constant names.

File: testing/utiltest.cc
Comment (twollgam) [reply testing/utiltest.cc:39]: @Baklap4 could you live with the comments (too ;-) )?

Comment (Baklap4) [reply testing/utiltest.cc:39]: Aight yea that's kinda fair :) 👍

**Comment by Baklap4** [comment]
Aight let's rebase this on top of latest master and get this tested 👍

**Comment by Baklap4** [comment]
Blocked by #864  (so might aswell rebase on top of that since that's gonna get merged to master once the ci is finished

File: game/configuration.cc
Comment (Baklap4) [inline game/configuration.cc:289]: Missing some braces here might aswell early continue to reduce nesting.
Or even better if such a functionality is available, already pre-filter the collection before looping as we only want strings which isn't the current selected value. in C# we use linq (in fluent syntax) a lot which does exactly that

File: game/hiscore.cc
Comment (Baklap4) [inline game/hiscore.cc:79]: Pre filter beforehand to eliminate the `if` and nesting and the forloop in general.
Can be done with a single line `std::copy_if` i assume since we're inserting in a new vector

File: game/screen_songs.cc
Comment (Baklap4) [inline game/screen_songs.cc:380]: I guess we can use range based loops and lambda's here something along the lines of:
```
auto stream = std::stringstream();
int n = 0;
const int maxLines = 10;

for (const auto& [track, scores] : scoresByTrack) {
    stream << track << ":\n";
    for (const auto& score : scores) {
        auto formatScore = [](const auto& score) {
            std::ostringstream ss;
            ss << std::right << std::setw(10) << score << " \t";
            return ss.str();
        };
        auto formatPlayer = [this](const auto& playerId) {
            const auto playerName = m_database.getPlayers().lookup(playerId);
            return std::left << std::setw(25) << (playerName ? *playerName : "Unknown player Id " + std::to_string(playerId));
        };
        stream << formatScore(score.score) << formatPlayer(score.playerid);
        if (score.unixtime.count()) {
            stream << " \t" << format(score.unixtime, datetimeFormat);
        }
        stream << "\n";
        if (++n == maxLines) {
            break;
        }
    }
    stream << "\n";
    if (n == maxLines) {
        break;
    }
}
```

File: game/songs.cc
Comment (Baklap4) [inline game/songs.cc:511]: This whole method can be simplified
```
void Songs::sortChange(Game& game, SortChange diff) {
    auto const numOrders = m_songOrders.size();

    switch (diff) {
        case SortChange::BACK:
            m_order = (m_order == 0) ? numOrders - 1 : m_order - 1;
            break;
        case SortChange::FORWARD:
            m_order = (m_order + 1) % numOrders;
            break;
        case SortChange::RESET:
            m_order = 0;
            break;
    }

    RestoreSel restore(*this);
    config["songs/sort-order"].ui() = m_order;

    if (m_order == 1 || m_order == 2 || m_order == 3 || m_order == 4 || m_order == 6) {
        UnicodeUtil::m_sortCollator->setStrength(config["game/case-sorting"].b() ? icu::Collator::TERTIARY : icu::Collator::SECONDARY);
    }

    sort_internal();
    writeConfig(game, false);
}

```
* Easier to read switch
* Changed second switch into an if statement
    * which can be even more simplified: `if(m_order >= 1 && m_order <=6)`

**Review by Baklap4 — CHANGES_REQUESTED** [review]

File: game/configuration.cc
Comment (twollgam) [reply game/configuration.cc:289]: Added the braces. The other point needs ranges, else we bloat the code.

File: game/songs.cc
Comment (twollgam) [reply game/songs.cc:511]: * I wrote a class Cycle that replaces the unsigned short. Looks much better.
* The second switch is nonsense and must be moved to the order classes. Advantage: Now the sequence can be changed without changing the switch or if or ...

File: game/screen_songs.cc
Comment (twollgam) [reply game/screen_songs.cc:380]: Yes I use lambdas now, but a little different as suggested.

Comment (Baklap4) [reply game/screen_songs.cc]: Instead of using stream.width(n) you can also inline it within the stream itself with `std::setw(n)` i think to remove the need of `stream.width(10);` and `stream.width(25);`
```
auto const scoreFormatter = [](auto& stream, auto const& score){
		stream.width(10);
		stream << std::right << score << " \t";
	};
	auto const playerFormatter = [this](auto& stream, auto const& playerId) {
		const auto player = m_database.getPlayers().lookup(playerId);

		stream.width(25);
		stream << std::left << player.value_or("Unknown player Id " + std::to_string(playerId));
	};
	auto const timeFormatter = [datetimeFormat](auto& stream, auto const& unixtime){
		if(unixtime.count()) {
			stream << " \t" << format(unixtime, datetimeFormat);
		}
	};
```

Comment (twollgam) [reply game/screen_songs.cc]: Using stream manipulators now.

**Comment by twollgam** [comment]
@Baklap4  GitHub shows a pending change request but I can't find it ?

**Review by Baklap4 — APPROVED** [review]

**Comment by twollgam** [comment]
@ooshlablu  can you test the current state once again before merging? Many thanks!

**Comment by ooshlablu** [comment]
Just tested the latest AppImage and debs with guitar/bass and vocals on Ubuntu 20.04 & 22.04. Everything looks good, it is definitely keeping track of the songs that are "Most Sung". When I sort by file time it picks up the newest songs correctly.
I didn't notice any issues with gameplay overall compared to master, so this LGTM :+1:

**Pull request merged**
